src/contracts/middleware/BLSRegistryCoordinatorWithIndices.sol

```diff
@@ -0,0 +1,325 @@
+        });
+
+        // record a stake update not bonding the operator at all (unbonded at 0), because they haven't served anything yet
+        serviceManager.recordFirstStakeUpdate(operator, 0);
+    }
+
+    function _deregisterOperatorWithCoordinator(address operator, bytes calldata quorumNumbers, BN254.G1Point memory pubkey, bytes32[] memory operatorIdsToSwap, uint32 globalOperatorListIndex) internal {
+        require(_operators[operator].status == OperatorStatus.REGISTERED, "BLSIndexRegistryCoordinator._deregisterOperatorWithCoordinator: operator is not registered");
+
+        // get the operatorId of the operator
+        bytes32 operatorId = _operators[operator].operatorId;
+        require(operatorId == pubkey.hashG1Point(), "BLSIndexRegistryCoordinator._deregisterOperatorWithCoordinator: operatorId does not match pubkey hash");
+
+        // get the quorumNumbers of the operator
+        uint192 quorumsToRemoveBitmap = uint192(BytesArrayBitmaps.orderedBytesArrayToBitmap_Yul(quorumNumbers));
+        uint256 operatorQuorumBitmapHistoryLengthMinusOne = _operatorIdToQuorumBitmapHistory[operatorId].length - 1;
+        uint192 quorumBitmapBeforeUpdate = _operatorIdToQuorumBitmapHistory[operatorId][operatorQuorumBitmapHistoryLengthMinusOne].quorumBitmap;
+        // check that the quorumNumbers of the operator matches the quorumNumbers passed in
+        require(
+            quorumBitmapBeforeUpdate & quorumsToRemoveBitmap == quorumsToRemoveBitmap,
+            "BLSIndexRegistryCoordinator._deregisterOperatorWithCoordinator: cannot deregister operator for quorums that it is not a part of"
+        );
+        // check if the operator is completely deregistering
+        bool completeDeregistration = quorumBitmapBeforeUpdate == quorumsToRemoveBitmap;
+        
+        // deregister the operator from the BLSPubkeyRegistry
+        blsPubkeyRegistry.deregisterOperator(operator, completeDeregistration, quorumNumbers, pubkey);
+
+        // deregister the operator from the StakeRegistry
+        stakeRegistry.deregisterOperator(operatorId, quorumNumbers);
+
+        // deregister the operator from the IndexRegistry
+        indexRegistry.deregisterOperator(operatorId, completeDeregistration, quorumNumbers, operatorIdsToSwap, globalOperatorListIndex);
+
+        // set the toBlockNumber of the operator's quorum bitmap update
+        _operatorIdToQuorumBitmapHistory[operatorId][operatorQuorumBitmapHistoryLengthMinusOne].nextUpdateBlockNumber = uint32(block.number);
+        
+        // if it is not a complete deregistration, add a new quorum bitmap update
+        if (!completeDeregistration) {
+            _operatorIdToQuorumBitmapHistory[operatorId].push(QuorumBitmapUpdate({
+                updateBlockNumber: uint32(block.number),
+                nextUpdateBlockNumber: 0,
+                quorumBitmap: quorumBitmapBeforeUpdate & ~quorumsToRemoveBitmap // this removes the quorumsToRemoveBitmap from the quorumBitmapBeforeUpdate
```

Comment:
this is a clever little operation. should we perhaps split this out into a separate line (for clarity) and have an associated event (for indexing/tracking)?

Reply:
don't need the event if we have one on the registries?